enumerables.rb

```diff
@@ -0,0 +1,188 @@
+      i += 1
+    end
+    self unless block_given?
+  end
+
+  # 2. Create my_each_with_index
+  def my_each_with_index
+    i = 0
+    while i < length
+      yield(self[i], i) if block_given?
+      i += 1
+    end
+    self unless block_given?
+  end
+
+  # 3. Create my_select
+  def my_select
+    arr = []
+    (0..length - 1).each do |i|
+      criteria = yield(self[i]) if block_given?
+      arr << self[i] if criteria
+    end
+    block_given? ? arr : self
+  end
+
+  # 4. Create my_all?
+  # Refactored my_all? & my_any? Helper method, to check Regex/Class/parameter passed in.
+  # Disable rubocop to avoid high-complexity alerts on helper method
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  def my_all_any_none_helper(sub_param, value)
+    # when main_param & block is not given.
+    return value ? true : false if sub_param.nil? && !block_given? # true if element is true
+    # when a Regex is passed as an argument
+    return sub_param.match?(value) if sub_param.class == Regexp
+    # when a class is passed as an argument
+    return value.class == sub_param if sub_param.is_a?(Class)
+    # when sub_param is given
+    return sub_param == value unless sub_param.nil?
+
+    yield(value) ? true : false
+  end
+  # rubocop:enable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+
+  # Refactoring my_all?, enlightened by mentor Rory Heiller
+  def my_all?(main_param = nil)
+    result = true
+    my_each do |element| # iterate over self
+      condition = my_all_any_none_helper(main_param, element) { |i| block_given? ? yield(i) : i }
+      result &&= condition # i equals each element here
+      break if result == false # once find a false, exit the loop
+    end
+    result
+  end
+
+  # 5. Create my_any?
+  def my_any?(main_param = nil)
+    result = false
+    my_each do |element| # iterate over self
+      condition = my_all_any_none_helper(main_param, element) { |i| block_given? ? yield(i) : i }
+      result ||= condition
+      break if result == true # once got a true, exit the loop
+    end
+    result
+  end
+
+  # 6. Create my_none?
+  def my_none?(main_param = nil)
+    result = false
+    my_each do |element| # iterate over self
+      condition = my_all_any_none_helper(main_param, element) { |i| block_given? ? yield(i) : i }
+      result ||= condition
+      break if result == true # once got a true, exit the loop
+    end
+    !result # inverse of my_any?
+  end
+
+  # 7. Create my_count
+  def my_count(param = nil)
+    counter = 0
+    return length if param.nil? && !block_given?
+
+    my_each do |i|
+      condition = yield(i) if block_given?
+      counter += 1 if condition || param == i
+    end
+    counter
+  end
+
+  # 8. Create my_map
+  #   def my_map
+  #     arr = []
+  #     each do |i|
+  #       condition = yield(i)
+  #       arr << condition
+  #     end
+  #     arr
+  #   end
+
+  # 9. Create my_inject
+  def my_inject(initial = nil, sym = nil)
```

Comment:
* [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
``` p (5..10).my_inject { |sum, n| sum + n } ``` != ```p (5..10).inject { |sum, n| sum + n }```
``` p (5..10).my_inject(1) { |product, n| product * n } != p (5..10).inject(1) { |product, n| product * n }```